### PR TITLE
DOC: update the pandas.DataFrame.apply docstring

### PR DESCRIPTION
**Pull request by mdeboc**
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [x] PR title is "DOC: update the <your-function-or-method> docstring"
- [x] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [x] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [x] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
################################################################################
###################### Docstring (pandas.DataFrame.apply) ######################
################################################################################

Apply a function along an axis of the `Series`.

Objects passed to the function are `Series` objects having as index
either the DataFrame's index (`axis=0`)
or the DataFrame's columns (`axis=1`).
If `result_type` is None, the final return type is the return
type of the applied function.
Otherwise, it depends on the `result_type` argument.

Parameters
----------
func : function
    Function to apply to each column or row.
axis : {0 or 'index', 1 or 'columns'}, default 0
    Axis along which the function is applied:

    * 0 or 'index': apply function to each column.
    * 1 or 'columns': apply function to each row.
broadcast : boolean, optional
    Only relevant for aggregation functions:

    * `False` or `None` : returns a `Series` whose length is the length
      of the index or the number of columns (based on the `axis`
      parameter)
    * `True` : results will be broadcast to the original shape
      of the frame, the original index and columns will be retained.

    .. deprecated:: 0.23.0
       This argument will be removed in a future version, replaced
       by result_type='broadcast'.

raw : boolean, default False
    * `False` : passes each row or column into a `Series` to the
      function.
    * `True` : the passed function will receive ndarray objects
      instead.
    If you are just applying a NumPy reduction function this will
    achieve much better performance.
reduce : boolean or None, default None
    Try to apply reduction procedures. If the `DataFrame` is empty,
    :meth:`apply` will use reduce to determine whether the result
    should be a `Series` or a `DataFrame`. If reduce is None (the
    default), :meth:`apply`'s return value will be guessed by calling
    func on an empty `Series`
    (note: while guessing, exceptions raised by `func` will be
    ignored).
    If reduce is True a `Series` will always be returned, and if
    `False` a `DataFrame` will always be returned.

    .. deprecated:: 0.23.0.
       This argument will be removed in a future version, replaced
       by result_type='reduce'.

result_type : {'expand', 'reduce', 'broadcast', None}
    These only act when `axis=1` (columns):

    * 'expand' : list-like results will be turned into columns.
    * 'reduce' : returns a `Series` if possible rather than expanding
      list-like results. This is the opposite of 'expand'.
    * 'broadcast' : results will be broadcast to the original shape
      of the `DataFrame`, the original index and columns will be
      retained.

    The default behaviour (`None`) depends on the return value of the
    applied function: list-like results will be returned as a `Series`
    of those. However if the apply function returns a `Series` these
    are expanded to columns.

    .. versionadded:: 0.23.0.

args : tuple
    Positional arguments to pass to function in addition to the
    array/series.
kwds :
    Additional keyword arguments to pass as keywords to the function.

Notes
-----
In the current implementation apply calls func twice on the
first column/row to decide whether it can take a fast or slow
code path. This can lead to unexpected behavior if func has
side-effects, as they will take effect twice for the first
column/row.

Examples
--------

We use this DataFrame to illustrate

>>> df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1,
...                   columns=['A', 'B', 'C'])
>>> df
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

Using a numpy universal function (in this case the same as
``np.sqrt(df)``):

>>> df.apply(np.sqrt)
     A         B         C
0  1.0  1.414214  1.732051
1  1.0  1.414214  1.732051
2  1.0  1.414214  1.732051
3  1.0  1.414214  1.732051
4  1.0  1.414214  1.732051
5  1.0  1.414214  1.732051

Using a reducing function on either axis

>>> df.apply(np.sum, axis=0)
A     6
B    12
C    18
dtype: int64

>>> df.apply(np.sum, axis=1)
0    6
1    6
2    6
3    6
4    6
5    6
dtype: int64

Retuning a list-like will result in a Series

>>> df.apply(lambda x: [1, 2], axis=1)
0    [1, 2]
1    [1, 2]
2    [1, 2]
3    [1, 2]
4    [1, 2]
5    [1, 2]
dtype: object

Passing result_type='expand' will expand list-like results
to columns of a Dataframe

>>> df.apply(lambda x: [1, 2], axis=1, result_type='expand')
   0  1
0  1  2
1  1  2
2  1  2
3  1  2
4  1  2
5  1  2

Returning a Series inside the function is similar to passing
``result_type='expand'``. The resulting column names
will be the Series index.

>>> df.apply(lambda x: pd.Series([1, 2], index=['foo', 'bar']), axis=1)
   foo  bar
0    1    2
1    1    2
2    1    2
3    1    2
4    1    2
5    1    2

Passing ``result_type='broadcast'`` will ensure the same shape
result, whether list-like or scalar is returned by the function,
and broadcast it along the axis. The resulting column names will
be the originals.

>>> df.apply(lambda x: [1, 2, 3], axis=1, result_type='broadcast')
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

See also
--------
DataFrame.applymap: For elementwise operations
DataFrame.aggregate: only perform aggregating type operations
DataFrame.transform: only perform transformating type operations

Returns
-------
applied : Series or DataFrame

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	Errors in parameters section
		Parameters {'kwds'} not documented
		Unknown parameters {'kwds :'}
		Parameter "raw" description should start with capital letter
		Parameter "kwds :" has no type
```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
 
     def apply(self, func, axis=0, broadcast=None, raw=False, reduce=None,
               result_type=None, args=(), **kwds):
-        """Applies function along an axis of the DataFrame.
+        """
+        Apply a function along an axis of the `Series`.
```
Comment (datapythonista) [inline pandas/core/frame.py]: Why did you change `DataFrame` by `Series`? The file is `frame.py` I think it's apply for `DataFrame`, isn't it?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
 
-        Objects passed to functions are Series objects having index
-        either the DataFrame's index (axis=0) or the columns (axis=1).
-        Final return type depends on the return type of the applied function,
-        or on the `result_type` argument.
+        Objects passed to the function are `Series` objects having as index
+        either the DataFrame's index (`axis=0`)
+        or the DataFrame's columns (`axis=1`).
+        If `result_type` is None, the final return type is the return
```
Comment (datapythonista) [inline pandas/core/frame.py]: `None` in backticks

File: pandas/core/frame.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: This is the dataframe docs, so Series -> DataFrame

(also it is not needed to put single backticks around Series and DataFrame, same for the rest of the docstring)

Comment (TomAugspurger) [inline pandas/core/frame.py]: \`Series\` -> DataFrame

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
               result_type=None, args=(), **kwds):
-        """Applies function along an axis of the DataFrame.
+        """
+        Apply a function along an axis of the `Series`.
 
-        Objects passed to functions are Series objects having index
-        either the DataFrame's index (axis=0) or the columns (axis=1).
-        Final return type depends on the return type of the applied function,
-        or on the `result_type` argument.
+        Objects passed to the function are `Series` objects having as index
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: "having as" -> "whose index is either"

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
+        either the DataFrame's index (`axis=0`)
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: double backticks around `axis=0` and `axis=1` since they're code snippets.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
         broadcast : boolean, optional
-            For aggregation functions, return object of same size with values
-            propagated
+            Only relevant for aggregation functions:
+
+            * `False` or `None` : returns a `Series` whose length is the length
+              of the index or the number of columns (based on the `axis`
+              parameter)
+            * `True` : results will be broadcast to the original shape
+              of the frame, the original index and columns will be retained.
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Since this keyword is deprecated, I am not sure it is that important to add such a detailed description (I would rather expand the explanation of the new keyword `result_type`)

File: pandas/core/frame.py [hunk 6]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            Function to apply to each column/row
+            Function to apply to each column or row.
         axis : {0 or 'index', 1 or 'columns'}, default 0
-            * 0 or 'index': apply function to each column
-            * 1 or 'columns': apply function to each row
+            Axis along which the function is applied:
+
+            * 0 or 'index': apply function to each column.
+            * 1 or 'columns': apply function to each row.
         broadcast : boolean, optional
```
Comment (datapythonista) [inline pandas/core/frame.py]: `bool` instead of `boolean` (also in the next 2)

File: pandas/core/frame.py [hunk 7]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
             .. deprecated:: 0.23.0
                This argument will be removed in a future version, replaced
                by result_type='broadcast'.
 
         raw : boolean, default False
-            If False, convert each row or column into a Series. If raw=True the
-            passed function will receive ndarray objects instead. If you are
-            just applying a NumPy reduction function this will achieve much
-            better performance
+            * `False` : passes each row or column into a `Series` to the
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: "passes each row or column into a `Series`" -> "passes each row or column as a `Series`"

Comment (TomAugspurger) [inline pandas/core/frame.py]: Maybe "passes each row or column as a Series to the function"

File: pandas/core/frame.py (hunk 4 above)
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: can you use double backticks around axis=0 (as ``` ``axis=0`` ```) (because it is a small code snippet)

File: pandas/core/frame.py [hunk 8]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
+              function.
+            * `True` : the passed function will receive ndarray objects
+              instead.
+            If you are just applying a NumPy reduction function this will
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: This applies to `raw=Ture` so keep it under the item formed by that `*`

File: pandas/core/frame.py [hunk 9]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            of those. However if the apply function returns a Series these
+            The default behaviour (`None`) depends on the return value of the
+            applied function: list-like results will be returned as a `Series`
+            of those. However if the apply function returns a `Series` these
             are expanded to columns.
 
-            .. versionadded:: 0.23.0
+            .. versionadded:: 0.23.0.
 
         args : tuple
```
Comment (datapythonista) [inline pandas/core/frame.py]: `args` is always a tuple, so we remove it. The validation script will probably complain, but the convention is to use `*args` and `**kwargs` (with the stars)

File: pandas/core/frame.py [hunk 10]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
             Positional arguments to pass to function in addition to the
-            array/series
-        Additional keyword arguments will be passed as keywords to the function
+            array/series.
+        kwds :
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: kwds -> `**kwds` (and the colon is not needed)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the PR! Added some first comments

File: pandas/core/frame.py [hunk 11]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            Try to apply reduction procedures. If the DataFrame is empty,
-            apply will use reduce to determine whether the result should be a
-            Series or a DataFrame. If reduce is None (the default), apply's
-            return value will be guessed by calling func an empty Series (note:
-            while guessing, exceptions raised by func will be ignored). If
-            reduce is True a Series will always be returned, and if False a
-            DataFrame will always be returned.
-
-            .. deprecated:: 0.23.0
+            Try to apply reduction procedures. If the `DataFrame` is empty,
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Don't need backticsk around Series / DataFrame

File: pandas/core/frame.py [hunk 12]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
+            :meth:`apply` will use reduce to determine whether the result
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I think this resolves to `pandas.apply`, which doesn't exist. SEe if you can rewreite to avoid the reference to apply

File: pandas/core/frame.py [hunk 13]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
+            should be a `Series` or a `DataFrame`. If reduce is None (the
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: backticsk around parameter names (`reduce`)

File: pandas/core/frame.py [hunk 14]
```diff
@@ -4818,66 +4818,82 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            .. versionadded:: 0.23.0
+            .. versionadded:: 0.23.0.
 
         args : tuple
             Positional arguments to pass to function in addition to the
-            array/series
-        Additional keyword arguments will be passed as keywords to the function
+            array/series.
+        kwds :
+            Additional keyword arguments to pass as keywords to the function.
```
Comment (datapythonista) [inline pandas/core/frame.py]: A personal opinion, but for me it'd be a bit more explicit to say "to pass as keyword arguments to `func`". I think saying "the function" should be clear enough for most people, but specifying which functions avoids any ambiguity. Also por `args`.

**Review by datapythonista — COMMENTED** [review]
Looks quite nice, added some comments.

File: pandas/core/frame.py (hunk 10 above)
Comment (TomAugspurger) [inline pandas/core/frame.py]: Just updated our docs on this, we're doing `**kwds`

File: pandas/core/frame.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/frame.py]: Ai, we were all doing the review at the same time :-)

File: pandas/core/frame.py (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/core/frame.py]: In this case we need to keep it as `args` and the 'tuple' type description is correct. This is a special case because it actually is a *keyword* with the name `args`, not `*args` in the signature

File: pandas/core/frame.py (hunk 1 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add PR review suggestions from @jorisvandenbossche into commit c593a70c5

File: pandas/core/frame.py (hunk 5 above)
Comment (mdeboc) [reply pandas/core/frame.py]: @jorisvandenbossche OK. But <<results will be broadcast to the original shape of the frame, the original index and columns will be retained.>> is already copied from the explanation of the new keyword `result_type`. So nothing to add in the paragraph of this new parameter. 

Shall I revert to the previous wording for `broadcast` parameter, _i.e._:
```
'''
broadcast : boolean, optional
    For aggregation functions, return object of same size with values
    propagated
'''
```
(...which is rather fuzzy),
or not ?

File: pandas/core/frame.py (hunk 7 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 8caf5f615

File: pandas/core/frame.py (hunk 4 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit c1214b42c

File: pandas/core/frame.py (hunk 10 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 2432ff573

File: pandas/core/frame.py (hunk 1 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit c593a70c5

File: pandas/core/frame.py (hunk 2 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add PR review suggestions from @datapythonista into commit da64c864c
Maybe ``result_type is None`` is better there, because it is a small code snippet.

File: pandas/core/frame.py (hunk 6 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit dbd90c1b8

File: pandas/core/frame.py (hunk 9 above)
Comment (mdeboc) [reply pandas/core/frame.py]: Ok, so I let the dot at the end.

File: pandas/core/frame.py (hunk 14 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 6cbadd8ed

File: pandas/core/frame.py (hunk 1 above)
Comment (mdeboc) [reply pandas/core/frame.py]: Yep. Sorry guys :)

File: pandas/core/frame.py (hunk 3 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add PR review suggestions from @TomAugspurger into commit eaf144156

File: pandas/core/frame.py (hunk 4 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit c1214b42c

File: pandas/core/frame.py (hunk 7 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 8caf5f615

File: pandas/core/frame.py (hunk 11 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit c593a70c5

File: pandas/core/frame.py (hunk 10 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 2432ff573

File: pandas/core/frame.py (hunk 13 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit da64c86
Maybe ``reduce is None`` is better there, because it is a small code snippet

File: pandas/core/frame.py (hunk 8 above)
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 3794d1989

File: pandas/core/frame.py (hunk 12 above)
Comment (mdeboc) [reply pandas/core/frame.py]: After running 
```bash
python make.py html --single pandas.DataFrame.apply
```
the :meth:`apply`  link resolves to `pandas.DataFrame.apply`

@TomAugspurger, if it is not the case for you, I shall revert to a wording more like: `apply()` or `DataFrame.apply()`... Let me know.

File: pandas/core/frame.py (hunk 1 above)
Comment (mdeboc) [reply pandas/core/frame.py]: @TomAugspurger I took into account your suggestion in the commit c593a70

File: pandas/core/frame.py [hunk 15]
```diff
@@ -4818,72 +4818,89 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            return value will be guessed by calling func an empty Series (note:
-            while guessing, exceptions raised by func will be ignored). If
-            reduce is True a Series will always be returned, and if False a
-            DataFrame will always be returned.
-
-            .. deprecated:: 0.23.0
+            :meth:`apply` will use `reduce` to determine whether the result
+            should be a Series or a DataFrame. If ``reduce is None`` (the
+            default), :meth:`apply`'s return value will be guessed by calling
+            `func` on an empty Series
```
Comment (jreback) [inline pandas/core/frame.py]: @jorisvandenbossche can you look at this formatting

File: pandas/core/frame.py [hunk 16]
```diff
@@ -4818,72 +4818,89 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-              list-like results. This is the opposite to 'expand'.
+            * 'reduce' : returns a Series if possible rather than expanding
+              list-like results. This is the opposite of 'expand'.
             * 'broadcast' : results will be broadcast to the original shape
-              of the frame, the original index & columns will be retained.
+              of the DataFrame, the original index and columns will be
+              retained.
 
-            The default behaviour (None) depends on the return value of the
+            The default behaviour (`None`) depends on the return value of the
```
Comment (jreback) [inline pandas/core/frame.py]: no parens here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 16 above)
Comment (mdeboc) [reply pandas/core/frame.py]: PR review suggestions from @jreback was added into commit c4da1ea72

File: pandas/core/frame.py (hunk 15 above)
Comment (jorisvandenbossche) [reply pandas/core/frame.py]: @mdeboc Although `` :meth:`apply` `` works correctly in sphinx rendering, I would not use the referencing syntax, as it is then just adding a reference to itself ? 
I would just do `` `apply` `` in this case

Comment (mdeboc) [reply pandas/core/frame.py]: @jorisvandenbossche I add this PR review suggestions into commit ee6919efd

File: pandas/core/frame.py [hunk 17]
```diff
@@ -4818,72 +4818,89 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-        Objects passed to functions are Series objects having index
-        either the DataFrame's index (axis=0) or the columns (axis=1).
-        Final return type depends on the return type of the applied function,
-        or on the `result_type` argument.
+        Objects passed to the function are Series objects whose index is
+        either the DataFrame's index (``axis=0``)
+        or the DataFrame's columns (``axis=1``).
+        If ``result_type is None``, the final return type is the return
+        type of the applied function.
+        Otherwise, it depends on the `result_type` argument.
```
[the remark this answers is not in the document]
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 23be01b34

File: pandas/core/frame.py [hunk 18]
```diff
@@ -4818,72 +4818,89 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
 
-        Objects passed to functions are Series objects having index
-        either the DataFrame's index (axis=0) or the columns (axis=1).
-        Final return type depends on the return type of the applied function,
-        or on the `result_type` argument.
+        Objects passed to the function are Series objects whose index is
+        either the DataFrame's index (``axis=0``)
+        or the DataFrame's columns (``axis=1``).
+        If ``result_type is None``, the final return type is the return
+        type of the applied function.
```
[the remark this answers is not in the document]
Comment (mdeboc) [reply pandas/core/frame.py]: add this PR review suggestions into commit 51ad0cb6c

File: pandas/core/frame.py [hunk 19]
```diff
@@ -4818,72 +4818,88 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
+              retained.
 
-            The default behaviour (None) depends on the return value of the
+            The default behaviour `None` depends on the return value of the
             applied function: list-like results will be returned as a Series
             of those. However if the apply function returns a Series these
             are expanded to columns.
 
-            .. versionadded:: 0.23.0
+            .. versionadded:: 0.23.0.
```
Comment (datapythonista) [inline pandas/core/frame.py]: There is a bug in the script, you don't need the period at the end.

File: pandas/core/frame.py (hunk 9 above)
Comment (datapythonista) [reply pandas/core/frame.py]: yes, my fault, sorry.

File: pandas/core/frame.py [hunk 20]
```diff
@@ -4818,72 +4818,88 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
 
         Notes
         -----
-        In the current implementation apply calls func twice on the
+        In the current implementation apply calls `func` twice on the
         first column/row to decide whether it can take a fast or slow
-        code path. This can lead to unexpected behavior if func has
+        code path. This can lead to unexpected behavior if `func` has
         side-effects, as they will take effect twice for the first
         column/row.
```
Comment (datapythonista) [inline pandas/core/frame.py]: Can you get rid of the "We use this DataFrame to illustrate" at the beginning of the examples? I don't think it adds any value.

Also, I'd personally prefer to create a smaller (e.g 3 rows, 2 cols) dataframe from a Python list, instead of all that numpy machinery. :) And if the example is with `sqrt`, I'd use numbers like `9` so the user can quickly see what `apply` does, which is the goal of this example.

**Review by datapythonista — COMMENTED** [review]
Good work, added some comments, mainly to parts of the original docstring that can be improved.

File: pandas/core/frame.py (hunk 19 above)
Comment (mdeboc) [reply pandas/core/frame.py]: Ok. see cf2f2d06b

File: pandas/core/frame.py (hunk 20 above)
Comment (mdeboc) [reply pandas/core/frame.py]: Yes, it's simpler. See d52ac81e7

Comment (mdeboc) [reply pandas/core/frame.py]: @datapythonista In commit 3160fb805, I suppress an unnecessary test, unintentionally introduced in previous commit (d52ac81)

File: pandas/core/frame.py [hunk 21]
```diff
@@ -4818,118 +4818,121 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            while guessing, exceptions raised by func will be ignored). If
-            reduce is True a Series will always be returned, and if False a
-            DataFrame will always be returned.
+            `apply` will use `reduce` to determine whether the result
+            should be a Series or a DataFrame. If ``reduce is None`` (the
+            default), `apply`'s return value will be guessed by calling
+            `func` on an empty Series
+            (note: while guessing, exceptions raised by `func` will be
+            ignored).
+            If ``reduce is True`` a Series will always be returned, and if
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: "reduce is True" -> "reduce=True" (same on line below and few lines above)

File: pandas/core/frame.py [hunk 22]
```diff
@@ -4818,118 +4818,121 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
-            better performance
-        reduce : boolean or None, default None
+        raw : bool, default False
+            * `False` : passes each row or column as a Series to the
+              function.
+            * `True` : the passed function will receive ndarray objects
+              instead.
+              If you are just applying a NumPy reduction function this will
+              achieve much better performance.
+        reduce : bool or `None`, default `None`
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: no backticks are needed inside the type description part

File: pandas/core/frame.py [hunk 23]
```diff
@@ -4950,36 +4951,27 @@ def apply(self, func, axis=0, broadcast=None, raw=False, reduce=None,
...
-        4  1  2  3
-        5  1  2  3
+        >>> df.apply(lambda x: [1, 2], axis=1, result_type='broadcast')
+           A  B
+        0  1  2
+        1  1  2
+        2  1  2
 
         See also
         --------
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Can you move the Returns and See Also sections to just after the Parameters section?

**Review by jorisvandenbossche — COMMENTED** [review]
Added some minor comments. Looking good for the rest!

File: pandas/core/frame.py (hunk 21 above)
Comment (mdeboc) [reply pandas/core/frame.py]: @jorisvandenbossche Changes in dfa0f5571, but don't forget that this keyword is deprecated, as you mentioned Saturday. :)

File: pandas/core/frame.py (hunk 23 above)
Comment (mdeboc) [reply pandas/core/frame.py]: Sure
changes in d757960db
@jorisvandenbossche  maybe the teams working with functions:  `combine`, `combine_first`, `applymap`, `join` and `round` should do the same...

**Review by TomAugspurger — APPROVED** [review]

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@mdeboc Thanks a lot!
